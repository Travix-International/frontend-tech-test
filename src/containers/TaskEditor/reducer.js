import {
  SAVE_TASK_FULFILLED,
  SAVE_TASK_PENDING,
  GET_TASK_FULFILLED,
  GET_TASK_PENDING,
  DELETE_TASK_FULFILLED,
  DELETE_TASK_PENDING,
} from './constants';

const reducer = (
  state = {
    task: { id: -1, title: '', description: '' },
    inProgress: false,
  },
  action
) => {
  switch (action.type) {
    case GET_TASK_PENDING:
      return { ...state, inProgress: true };
    case GET_TASK_FULFILLED:
      return {
        ...state,
        task: action.payload.task,
        inProgress: false,
      };
    case SAVE_TASK_PENDING:
      return { ...state, inProgress: true };
    case SAVE_TASK_FULFILLED:
      return {
        ...state,
        inProgress: false,
        task: action.payload.task,
      };

    case DELETE_TASK_PENDING:
      return { ...state, inProgress: true };
    case DELETE_TASK_FULFILLED:
      return { ...state, inProgress: false };
    default:
      return state;
  }
};

export default reducer;
