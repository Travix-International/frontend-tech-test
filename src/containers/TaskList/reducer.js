import { FETCH_TASKS_FULFILLED, FETCH_TASKS_PENDING } from './constants';

const reducer = (state = { fetching: true }, action) => {
  switch (action.type) {
    case FETCH_TASKS_PENDING:
      return { ...state, fetching: true };
    case FETCH_TASKS_FULFILLED:
      return {
        ...state,
        fetching: false,
        tasks: action.payload.tasks,
        pageNumber: action.payload.pageNumber,
        totalRecords: action.payload.totalRecords,
      };
    default:
      return state;
  }
};

export default reducer;
