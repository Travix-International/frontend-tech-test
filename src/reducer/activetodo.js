import { SET_ACTIVE_TODO } from '../constants/actions.js';

export default function activeTodo(state = null, action) {
  switch (action.type) {
    case SET_ACTIVE_TODO:
      return action.payload;
    default:
      return state;
  }
}