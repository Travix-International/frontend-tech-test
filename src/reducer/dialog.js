import {
  SHOW_DIALOG,
  HIDE_DIALOG,
  CHANGE_DIALOG_VIEW,
  CHANGE_DIALOG_FIELD
} from '../constants/actions.js';

const initialState = {
  isOpened: false,
  isTodoChanges: false,
  form: {
    title: '',
    description: '',
    subtasks: [],
    tags: []
  }
};

export default function dialog(state = initialState, action) {
  switch (action.type) {
    case SHOW_DIALOG:
      return { ...initialState, isOpened: true };
    case HIDE_DIALOG:
      return { ...state, isOpened: false };
    case CHANGE_DIALOG_VIEW:
      return { ...state, isTodoChanges: action.payload };
    case CHANGE_DIALOG_FIELD:
      return { ...state, form: { ...state.form, ...action.payload } };
    default:
      return state;
  }
}