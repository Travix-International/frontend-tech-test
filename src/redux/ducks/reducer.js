import { combineReducers } from 'redux-immutable'

/* Resources reducers imports */
import todos from './todos'
import settings from './settings'

/* UI reducers imports */

/* Resources reducer */
const resources = combineReducers({
  todos,
  settings,
})

/* UI Reducer */

/* Root Reducer */
const rootReducer = combineReducers({ resources })

export default rootReducer
