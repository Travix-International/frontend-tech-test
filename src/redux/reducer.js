import { reducer as formReducer } from 'redux-form/immutable'
import { combineReducers } from 'redux-immutable'
import { Map } from 'immutable'

/* Resources reducers imports */
import todos, { ADD_TODO } from 'redux/ducks/todos'
import settings from 'redux/ducks/settings'

/* Utils */
export const clearReduxForm = state => (
  state
    .set('values', Map({}))
    .set(
      'fields',
      state.get('fields').map(field => field.set('touched', false)),
    )
)

/* UI reducers imports */

/* Resources reducer */
const resources = combineReducers({
  todos,
  settings,
})

/* Input reducer */
const form = formReducer.plugin({
  addTodo: (state, action) => {
    switch (action.type) {
      case ADD_TODO.SUCCESS:
        return clearReduxForm(state)
      default:
        return state
    }
  },
})

/* UI Reducer */

/* Root Reducer */
const rootReducer = combineReducers({
  resources,
  form,
})

export default rootReducer
