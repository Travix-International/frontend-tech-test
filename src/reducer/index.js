import { combineReducers } from 'redux';

import todos from './todos.js';
import activeTodo from './activetodo.js';
import dialog from './dialog.js';
import tags from './tags.js';

export default combineReducers({
  todos,
  activeTodo,
  dialog,
  tags
});