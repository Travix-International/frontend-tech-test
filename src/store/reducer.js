import { combineReducers } from 'redux';
import todoReducer from '../containers/app/app.reducer';
export default combineReducers({
    todo: todoReducer,
});
