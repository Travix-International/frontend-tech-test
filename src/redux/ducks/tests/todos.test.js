import { fromJS, is, Map } from 'immutable'
import { put, call } from 'redux-saga/effects'
import * as matchers from 'jest-immutable-matchers'

import reducer, {
  ADD_TODO,
  COMPLETE_TODO,
  DELETE_TODO,
  LOAD_TODOS,
  addTodo,
  completeTodo,
  deleteTodo,
  loadTodos,
  createTodoSaga,
  completeTodoSaga,
  deleteTodoSaga,
  loadTodosSaga,
} from 'redux/ducks/todos'

describe('todos duck', () => {
  beforeAll(() => {
    jest.addMatchers(matchers)
  })

  describe('actions', () => {
    it('completeTodo', () => {
      const fixture = { id: '1337' }
      const expected = {
        type: COMPLETE_TODO.ACTION,
        payload: fixture,
      }

      expect(completeTodo(fixture)).toEqual(expected)
    })

    it('deleteTodo', () => {
      const fixture = { id: '1337' }
      const expected = {
        type: DELETE_TODO.ACTION,
        payload: fixture,
      }

      expect(deleteTodo(fixture)).toEqual(expected)
    })

    it('addTodo', () => {
      const fixture = { id: '1337' }
      const expected = {
        type: ADD_TODO.ACTION,
        payload: fixture,
      }

      expect(addTodo(fixture)).toEqual(expected)
    })

    it('loadTodos', () => {
      const expected = {
        type: LOAD_TODOS.ACTION,
      }

      expect(loadTodos()).toEqual(expected)
    })
  })

  describe('reducer', () => {
    const fixture1 = { id: '1' }
    const fixture2 = { id: '2' }
    const fixture3 = { id: '3' }

    let initialState

    beforeAll(() => {
      initialState = fromJS({
        items: [
          { title: 'Create a To-do app', description: 'Use all you know about react/redux to do so!', id: '1', done: false },
          { title: 'Create a Pull Request', description: 'Submit yoursolution for review!', id: '2', done: true },
          { title: 'Eat!', description: 'It\'s apparently quite important', id: '3', done: false },
          { title: 'Rest!', description: 'Like eating, it seems like this is also really important', id: '4', done: false },
          { title: 'Have social life!', description: 'Ok, now we\'re just being ridiculous...', id: '5', done: true },
          { title: 'Work!', description: 'THERE\'S NOT ENOUGH TIME!', id: '6', done: false },
        ],
      })
    })

    it('should return initial immutable state', () => {
      expect(reducer(undefined, {})).toBeImmutableMap()
    })

    it('should handle addTodo', () => {
      const newTodo = Map({ title: 'I am new!', description: 'Just created!', id: 7, done: false })
      const getExpected = fixture => (
        initialState.set('items', initialState.get('items').insert(0, fixture))
      )

      expect(is(reducer(initialState, {
        type: ADD_TODO.SUCCESS,
        payload: newTodo,
      }), getExpected(newTodo))).toEqual(true)
    })

    it('should handle completeTodo', () => {
      const getExpected = fixture => (
        initialState.set('items', initialState.get('items').map((item) => {
          if (item.get('id') !== fixture.get('id')) { return item }
          return item.set('done', !item.get('done'))
        }))
      )

      expect(reducer(initialState, {
        type: COMPLETE_TODO.SUCCESS,
        payload: Map(fixture1),
      })).toEqual(getExpected(Map(fixture1)))
      expect(reducer(initialState, {
        type: COMPLETE_TODO.SUCCESS,
        payload: Map(fixture2),
      })).toEqual(getExpected(Map(fixture2)))
      expect(reducer(initialState, {
        type: COMPLETE_TODO.SUCCESS,
        payload: Map(fixture3),
      })).toEqual(getExpected(Map(fixture3)))
    })

    it('should handle deleteTodo', () => {
      const getExpected = fixture => (
        initialState.set('items', initialState.get('items').filter(item => item.get('id') !== fixture.get('id')))
      )

      expect(reducer(initialState, {
        type: DELETE_TODO.SUCCESS,
        payload: Map(fixture1),
      })).toEqual(getExpected(Map(fixture1)))
      expect(reducer(initialState, {
        type: DELETE_TODO.SUCCESS,
        payload: Map(fixture2),
      })).toEqual(getExpected(Map(fixture2)))
      expect(reducer(initialState, {
        type: DELETE_TODO.SUCCESS,
        payload: Map(fixture3),
      })).toEqual(getExpected(Map(fixture3)))
    })

    it('should handle loadTodos', () => {
      const loadedTodos = fromJS([
        { title: 'Get the triforce of courage', description: 'Need it for kicking ass', id: 'hyrule-123', done: true },
        { title: 'Get the triforce of wisdow', description: 'Need it for solving puzzles', id: 'hyrule-124', done: true },
        { title: 'Get the triforce of power', description: 'Take it from that pig-face!', id: 'hyrule-125', done: false },
      ])
      const getExpected = fixture => initialState.set('items', fixture)

      expect(reducer(initialState, {
        type: LOAD_TODOS.SUCCESS,
        payload: loadedTodos,
      })).toEqual(getExpected(loadedTodos))
    })
  })

  describe('sagas', () => {
    describe('createTodoSaga', () => {
      const fixture = { payload: fromJS({}) }
      let generator

      beforeEach(() => {
        generator = createTodoSaga(fixture)
      })

      it('should call an api function and dispatch ADD_TODO.SUCCESS on success', () => {
        const callDescriptor = generator.next().value
        const response = { name: 'Link' }
        const putDescriptor = generator.next(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: ADD_TODO.SUCCESS, payload: fromJS(response) }))
      })

      it('should call ADD_TODO.ERROR on error', () => {
        const callDescriptor = generator.next().value
        const response = new Error('Ganon wins!')
        const putDescriptor = generator.throw(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: ADD_TODO.ERROR, payload: { error: response } }))
      })
    })

    describe('completeTodoSaga', () => {
      const fixture = { payload: fromJS({}) }
      let generator

      beforeEach(() => {
        generator = completeTodoSaga(fixture)
      })

      it('should call an api function and dispatch COMPLETE_TODO.SUCCESS on success', () => {
        const callDescriptor = generator.next().value
        const response = { name: 'Zelda' }
        const putDescriptor = generator.next(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: COMPLETE_TODO.SUCCESS, payload: fromJS(response) }))
      })

      it('should call COMPLETE_TODO.ERROR on error', () => {
        const callDescriptor = generator.next().value
        const response = new Error('Ganon wins!')
        const putDescriptor = generator.throw(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: COMPLETE_TODO.ERROR, payload: { error: response } }))
      })
    })

    describe('deleteTodoSaga', () => {
      const fixture = { payload: fromJS({}) }
      let generator

      beforeEach(() => {
        generator = deleteTodoSaga(fixture)
      })

      it('should call an api function and dispatch DELETE_TODO.SUCCESS on success', () => {
        const callDescriptor = generator.next().value
        const response = { name: 'Ganon' }
        const putDescriptor = generator.next(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: DELETE_TODO.SUCCESS, payload: fromJS(response) }))
      })

      it('should call DELETE_TODO.ERROR on error', () => {
        const callDescriptor = generator.next().value
        const response = new Error('Ganon wins!')
        const putDescriptor = generator.throw(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: DELETE_TODO.ERROR, payload: { error: response } }))
      })
    })

    describe('loadTodosSaga', () => {
      const fixture = { payload: fromJS({}) }
      let generator

      beforeEach(() => {
        generator = loadTodosSaga(fixture)
      })

      it('should call an api function and dispatch LOAD_TODOS.SUCCESS on success', () => {
        const callDescriptor = generator.next().value
        const response = { name: 'Ingo' }
        const putDescriptor = generator.next(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: LOAD_TODOS.SUCCESS, payload: fromJS(response) }))
      })

      it('should call LOAD_TODOS.ERROR on error', () => {
        const callDescriptor = generator.next().value
        const response = new Error('Ganon wins!')
        const putDescriptor = generator.throw(response).value

        expect(callDescriptor).toMatchSnapshot()
        // eslint-disable-next-line
        expect(putDescriptor).toEqual(put({ type: LOAD_TODOS.ERROR, payload: { error: response } }))
      })
    })
  })
})
