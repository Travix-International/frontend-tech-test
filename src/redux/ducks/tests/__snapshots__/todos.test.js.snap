// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos duck sagas completeTodoSaga should call COMPLETE_TODO.ERROR on error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas completeTodoSaga should call an api function and dispatch COMPLETE_TODO.SUCCESS on success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas createTodoSaga should call ADD_TODO.ERROR on error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas createTodoSaga should call an api function and dispatch ADD_TODO.SUCCESS on success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas deleteTodoSaga should call DELETE_TODO.ERROR on error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas deleteTodoSaga should call an api function and dispatch DELETE_TODO.SUCCESS on success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas loadTodosSaga should call LOAD_TODOS.ERROR on error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`todos duck sagas loadTodosSaga should call an api function and dispatch LOAD_TODOS.SUCCESS on success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
